src/troute-network/troute/HYFeaturesNetwork.py

```diff
@@ -287,114 +360,134 @@ def read_geo_file(self,):
-                                                                    'waterbody_toid': 'toid'})
-            self._waterbody_df.set_index("wb-id", inplace=True)
-            self._waterbody_df = self.waterbody_dataframe.sort_index()
-
-            self._waterbody_types_df = pd.DataFrame(self.waterbody_dataframe['reservoir_type'])
-            self._waterbody_df.drop('reservoir_type', axis=1, inplace=True)
+            self._waterbody_connections = {}
+            self._waterbody_type_specified = False
+            self._link_lake_crosswalk = None
+
+    def preprocess_data_assimilation(self, network):
+        if not network.empty:
+            gages_df = network[['id','hl_uri','hydroseq']].drop_duplicates()
+            # clear out missing values
+            gages_df = gages_df[~gages_df['hl_uri'].isnull()]
+            gages_df = gages_df[~gages_df['hydroseq'].isnull()]
+            # make 'id' an integer
+            gages_df['id'] = gages_df['id'].str.split('-',expand=True).loc[:,1].astype(float).astype(int)
+            # split the hl_uri column into type and value
+            gages_df[['type','value']] = gages_df.hl_uri.str.split('-',expand=True,n=1)
+            # filter for 'Gages' only
+            gages_df = gages_df[gages_df['type'].isin(['Gages','NID'])]
+            # assign only the furthest downstream segment ID to any given gage
+            gages_df = gages_df.sort_values(['value','hydroseq']).groupby('value').last().reset_index()
+            # transform dataframe into a dictionary
+            self._gages = gages_df[['id','value']].rename(columns={'value': 'gages'}).set_index('id').to_dict()
+
+            # Create lake_gage crosswalk dataframes:
+            link_gage_df = (
+                pd.DataFrame.from_dict(self._gages)
+                .reset_index()
+                .rename(columns={'index': 'link'})
+                .set_index('link')
+                )
+            lake_gage_df = (
+                pd.DataFrame(self.waterbody_connections.items(),
+                             columns = ['link', 'lake_id'])
+                             .set_index('link')
+                             .join(link_gage_df)
+                             .reset_index()[['lake_id','gages']]
+                )
+            lake_gage_df = lake_gage_df[~lake_gage_df['gages'].isnull()]
+            #################################
+            #FIXME: temporary solution, this makes sure each lake only has one gage, and that
+            # it is the gage furthest downstream
+            lake_ids = lake_gage_df.lake_id.unique()
+            lake_outflow_gage_ids = []
+
+            for i in lake_ids:
+                temp_df = lake_gage_df[lake_gage_df['lake_id']==i]
+                lake_outflow_gage_ids.append(
+                    gages_df[gages_df['value']
+                             .isin(temp_df.gages)]
+                             .sort_values('hydroseq')
```

Comment:
Does this sort_values() by 'hydroseq' guarantee to find a gage that is furthest downstream among, if any, multiple gages linked to the same watebody?